### PR TITLE
12.0 credit memo checks 3093579 mpv

### PR DESCRIPTION
**Pull request by mpv-odoo**
Description
Adds credit notes into checks for the linked payments. 

Link to task: [3093579](https://www.odoo.com/web#model=project.task&id=3093579)

### All Submissions:

* [ ] My commit respects the [Odoo commit guideline](https://www.odoo.com/documentation/15.0/developer/misc/other/guidelines.html#git)
* [ ] My commit message respects the [commit template](https://github.com/odoo-ps/psbe-process/wiki/Commits-message-guidelines#template)
* [ ] I have used pre-commit
* [ ] The PR contains **only** my modification and **no other external** commit

### Sh/Runbot:

* [ ] The commits pass test and the branch is green
* [ ] Unit tests have been implemented / standard ones rewritten
* [ ] The Staging is ISO-Prod and will contain only this dev

### Upgrade:

* [ ] The data affected (*if any*) by the changes has been migrated 

### Maintenance:

* [ ] The versions of the modules affected by the changes have been bumped in their manifests

### CODE REVIEW

File: caldoor_account_vendor_discounts/report/print_check.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
                 <th>Discount</th>
             </xpath>
             <xpath expr="//td[4]/t" position="replace">
-                <t t-esc="stub_line['discount']"/>
+                <t t-esc="stub_line.get('discount', '')"/>
```
Comment (kak-odoo) [inline caldoor_account_vendor_discounts/report/print_check.xml:10]: This should probably just us `t-if` instead

File: caldoor_account_vendor_discounts/report/print_check.xml [hunk 2]
```diff
@@ -7,7 +7,7 @@
                 <th>Discount</th>
             </xpath>
             <xpath expr="//td[4]/t" position="replace">
```
Comment (kak-odoo) [inline caldoor_account_vendor_discounts/report/print_check.xml:9]: I know you didn't write this, but  this sort of direct indexing is bad, it's hard to follow and almost impossible to maintain, we should change it

File: caldoor_checks/report/print_check.py [hunk 3]
```diff
@@ -1,11 +1,16 @@
 # -*- coding: utf-8 -*-
+import json
```
Comment (kak-odoo) [inline caldoor_checks/report/print_check.py:2]: add a new line between the comment and the beginning of imports

File: caldoor_checks/report/print_check.py [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
 import logging
 from num2words import num2words
+from datetime import datetime
 from odoo import api, fields, models, tools, _
+from odoo.addons.account_check_printing.models.account_payment import INV_LINES_PER_STUB
+from odoo.tools.misc import formatLang, format_date
```
Comment (kak-odoo) [inline caldoor_checks/report/print_check.py:8]: Use correct import order: https://www.odoo.com/documentation/15.0/contributing/development/coding_guidelines.html#imports

File: caldoor_checks/report/print_check.py [hunk 5]
```diff
@@ -77,4 +82,86 @@ def _check_make_stub_line(self, invoice):
...
+                    'invoice_date': format_date(self.env, credit_note.get('date')),
+                    'number': 'Credit Note: ' + credit_note.get('name', ''),
+                    'reference': credit_note.get('ref', '')
+                })
+            
+            if credits:
+                stub_lines += [{'header': True, 'name': "Refunds"}]
+                stub_lines += [self._check_make_stub_line(inv) for inv in credits]
+
+        ''' END CHANGES '''
```
Comment (kak-odoo) [inline caldoor_checks/report/print_check.py:148]: is all of this actually different, or at there just a bunch of smaller changes in this block?

**Review by kak-odoo — CHANGES_REQUESTED** [review]

File: caldoor_account_vendor_discounts/report/print_check.xml (hunk 2 above)
Comment (mpv-odoo) [reply caldoor_account_vendor_discounts/report/print_check.xml:9]: So I did write this a while ago, but only because it was the only way to grab that t tag. The only unique thing about the tag was its `t-esc` but I can't select it based on that since there's no quote escape character in xpath (all the ones I tried didn't work). The alternative would be to essentially override the entire t-foreach, but I wasn't sure if that was better or worse. 

Here's the source code if you want to take a look at it: https://github.com/odoo/enterprise/blob/12.0/l10n_us_check_printing/report/print_check.xml#L54

File: caldoor_checks/report/print_check.py (hunk 5 above)
Comment (mpv-odoo) [reply caldoor_checks/report/print_check.py:148]: Yeah it's all different.

File: caldoor_account_vendor_discounts/report/print_check.xml (hunk 2 above)
Comment (kak-odoo) [reply caldoor_account_vendor_discounts/report/print_check.xml:9]: You will need to handle it at the xml level, so `//t[@t-esc=&quot;stub_line['amount_residual']&quot;]` should work

**Review by kak-odoo — CHANGES_REQUESTED** [review]

File: caldoor_checks/report/print_check.py [hunk 6]
```diff
@@ -77,4 +83,86 @@ def _check_make_stub_line(self, invoice):
...
+            payments = json.loads(invoice.payments_widget).get('content')
+            credit_notes = [payment for payment in payments if not payment.get('account_payment_id', False)]
+            credit_amount = 0
+            '''
+            If there is a credit note, we need to change the payment amount to
+            be the "full amount" of the 
+            '''
+            if credit_notes:
+                credit_amount = sum([credit_note.get('amount') for credit_note in credit_notes])
+                amount_paid = float(invoice_stub_line.get('amount_paid').replace('$', '').replace(' ', ''))
```
Comment (kak-odoo) [inline caldoor_checks/report/print_check.py:127]: Would using re make this cleaner? Also why are we stripping spaces?

File: caldoor_checks/report/print_check.py [hunk 7]
```diff
@@ -77,4 +83,86 @@ def _check_make_stub_line(self, invoice):
...
+                amount_paid += credit_amount
+                invoice_stub_line['amount_paid'] = formatLang(self.env, amount_paid, currency_obj=self.currency_id)
+            stub_lines.append(invoice_stub_line)
+
+            for credit_note in credit_notes:
+                credit_amount = '(' + formatLang(self.env, credit_note.get('amount'), currency_obj=self.currency_id) + ')'
```
Comment (kak-odoo) [inline caldoor_checks/report/print_check.py:133]: use string formatting

File: caldoor_checks/report/print_check.py [hunk 8]
```diff
@@ -77,4 +83,86 @@ def _check_make_stub_line(self, invoice):
...
+            while i < len(stub_lines):
+                # Make sure we don't start the credit section at the end of a page
+                if len(stub_lines) >= i + INV_LINES_PER_STUB and stub_lines[i + INV_LINES_PER_STUB - 1].get('header'):
+                    num_stub_lines = INV_LINES_PER_STUB - 1 or INV_LINES_PER_STUB
+                else:
+                    num_stub_lines = INV_LINES_PER_STUB
+                stub_pages.append(stub_lines[i:i + num_stub_lines])
+                i += num_stub_lines
+
+        return stub_pages
```
Comment (kak-odoo) [inline caldoor_checks/report/print_check.py:168]: add a new line at the end of files

**Review by kak-odoo — CHANGES_REQUESTED** [review]

**Review by bmu-odoo — APPROVED** [review]
